inference/conflict.go

```diff
@@ -19,148 +19,177 @@ import (
 	"go/token"
 
 	"go.uber.org/nilaway/annotation"
+	"go.uber.org/nilaway/util"
 	"golang.org/x/tools/go/analysis"
 )
 
-// For concise error reporting, conflicts are grouped by their nilness root cause, i.e., by their producers. Thus, the producer
-// forms the 'key' to identify a conflict group.
+type conflict struct {
+	position token.Pos // stores position where the error should be reported
+	expr     string    // stores expression that is overcontrained (i.e., expected to be nonnil, but found nilable)
+	nilFlow  nilFlow   // stores nil flow from source to dereference point
+}
 
-// groupKey stores the type of the key representing a conflict group.
-// (Note that the type is currently set to any since the producer types of singleAssertionConflict and overconstrainedConflict
-// are not the same. Consolidating the two types of conflicts will help solve this problem (tracked in .)
-type groupKey any
+type nilFlow struct {
+	nilPath    []node // stores nil path of the flow from nilable source to conflict point
+	nonnilPath []node // stores non-nil path of the flow from conflict point to dereference point
+}
 
-// conflictGrouping stores groups of conflicts represented by their keys in a map. It also provides methods to add conflicts
-// to the groups and to generate diagnostics from the groups.
-type conflictGrouping struct {
-	groupMap map[groupKey][]conflict
+type node struct {
+	position token.Position
+	reason   string
 }
 
-// newConflictGrouping constructs and returns a pointer to conflictGrouping.
-func newConflictGrouping() *conflictGrouping {
-	return &conflictGrouping{
-		groupMap: make(map[groupKey][]conflict),
+func newNode(p annotation.Prestring) node {
+	nodeObj := node{}
+	if l, ok := p.(annotation.LocatedPrestring); ok {
+		nodeObj.position = l.Location
+		nodeObj.reason = l.Contained.String()
+	} else if p != nil {
+		nodeObj.reason = p.String()
 	}
+	return nodeObj
 }
 
-// addConflict adds a conflict to a conflict group identified by the conflict's key.
-func (t *conflictGrouping) addConflict(conflict conflict) {
-	key := conflict.key()
-	t.groupMap[key] = append(t.groupMap[key], conflict)
+func (n *node) String() string {
+	posStr := "<no pos info>"
+	reasonStr := "<no reason info>"
+	if n.position.IsValid() {
```

Comment:
Is this one of those cases where a method supports nilable receiver? (since I see from the function above that `n` could just be `node{}`)

Reply:
Not really. `newNode` is returning a struct, not a pointer, so the receiver `n` here of `String()` is never nil. Also, `n.position`, which is of type `token.Position`, is also a struct, not a pointer, so we can safely invoke the method `IsValid()` on it.

Reply:
Ah, right, too used to Java reference types. I did mean `n.position` would be `nil` by default, but, you are right, since it is a struct type and not a pointer to struct, it will just be an empty struct itself...